src/python-common/ceph/rgw/rgwam_core.py

```diff
@@ -770,22 +782,43 @@ def get_realms_info(self):
         realms_info = []
         for realm_name in self.realm_op().list():
             realm = self.get_realm(realm_name)
-            master_zone_inf = self.period_op().get_master_zone(realm)
-            zone_ep = self.period_op().get_master_zone_ep(realm)
-            if master_zone_inf and 'system_key' in master_zone_inf:
-                access_key = master_zone_inf['system_key']['access_key']
-                secret = master_zone_inf['system_key']['secret_key']
-            else:
-                access_key = ''
-                secret = ''
-            realms_info.append({"realm_name": realm_name,
-                                "realm_id": realm.id,
-                                "master_zone_id": master_zone_inf['id'] if master_zone_inf else '',
-                                "endpoint": zone_ep[0] if zone_ep else None,
-                                "access_key": access_key,
-                                "secret": secret})
+            realm_period = self.period_op().get(realm)
+            master_zone_id = realm_period['master_zone']
+            master_zone_name = self.get_master_zone_name(realm_period, master_zone_id)
+            local_zone_list = self.zone_op().list()
+
+            # Only consider the realm if master_zone_name is in the local zone list
+            if master_zone_name in local_zone_list:
+                master_zone_inf = self.zone_op().get(EntityID(master_zone_id))
+                zone_ep = self.period_op().get_master_zone_ep(realm)
+
+                if master_zone_inf and 'system_key' in master_zone_inf:
+                    access_key = master_zone_inf['system_key']['access_key']
+                    secret = master_zone_inf['system_key']['secret_key']
+                else:
+                    access_key = ''
+                    secret = ''
+
+                realms_info.append({
+                    "realm_name": realm_name,
+                    "realm_id": realm.id,
+                    "master_zone_id": master_zone_inf['id'] if master_zone_inf else '',
+                    "endpoint": zone_ep[0] if zone_ep else None,
+                    "access_key": access_key,
+                    "secret": secret
+                })
+
```

Comment:
Part of my issues with this code is that we tend to design our code after methods and use plain/dumb dicts for data holding, rather than using solid (and SOLID) data structures with methods (real/thick classes), and we spend lots of lines of code parsing/iterating/transforming ~~data structures~~ dicts to extract and accommodate information. This results in code hard to read/follow/swallow.

If we implemented classes (data + methods) after existing entities (`RGWDaemon`, `Realm`, `Zonegroup`, `Zone`) and specific methods in those classes to interact and provide the information needed, the services/methods would have been much simpler... I laugh (bitterly) every time I hear that OOP is dead...

Reply:
Updated, thanks